### DISCUSSION TITLE
How do I expand "rust-analyzer's "expand macros" command?

### DISCUSSION THREAD

**Discussion by dpc** [kakoune-lsp/kakoune-lsp#635] [Q&A] [answered]
It was pointed out to me that:

> Usually, when writing and debugging macros and I wonder what exactly the macro is doing, I use an excellent rust-analyzer command: Rust analyzer: Expand macro recursively.

However, I have no idea how to do it in `kak-lsp`. I have a `,l` user-layer configured for `kak-lsp` and do not see it in the list. "Show code actions for current position" doesn't seem to do anything.

I'm not sure if "rust-analyzer command" is the same as "action" anyway (I guess it isn't, because I can see, list and use actions just fine).

**Comment by sclu1034** [comment, marked as answer]
Commands are indeed not the same as Code Actions. And since those commands are custom to rust-analyzer and not part of the LSP spec, each one has to be manually handled/implemented in kak-lsp. And that's not been done for `rust-analyzer/expandMacro`, yet, so as of now, the answer is "you can't".

**Comment by krobelus** [comment]
Right, this is a [rust-analyzer LSP extension](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#expand-macro). Apparently they have like 20 of those..
Added in 8a14601c (Implement rust-analyzer-expand-macro for basic support for rust-analyzer/expandMacro, 2022-07-29)

**Reply by krobelus** [reply]
I guess the info box is enough, unless you want to use go-to-definition

**Reply by dpc** [reply]
Oh, wow. That was fast. Let me `:plug-update` and see if it works. :)

**Reply by dpc** [reply]
I see it available, and it shows the output in the info box for 0.2s and then dissapears. I think something like that (info box closing after a short time) happens in other places and might be something about my config. :/

**Reply by krobelus** [reply]
Yeah that sounds like a bad NormalIdle hook in your config.

**Reply by dpc** [reply]
Found it - it was my lame auto-save on leaving Insert mode hook,  I think. I guess this works, however the info box is kind of limiting for macros that expand to a lot of text.

**Reply by krobelus** [reply]
if you put the autosave in a draft context it should not interfere with any UI

we could define a callback like `lsp-show-expand-macro` or just hardcode to open a scratch buffer, or insert the expansion in the current buffer..

**Reply by sclu1034** [reply]
I think just always using a scratch buffer would be fine here.

**Reply by dpc** [reply]
I kind of like the expansion in the info buffer. It's handy. But eventually it will not fit, that's all. Not sure if it should just always go to scratch buffer, or when its too large, or maybe there should be two commands, or an option. :shrug:

**Reply by dpc** [reply]
I guess always scratch buffer is going to be handy too.
